pkg/policyreport/eraser.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/kyverno/kyverno/pkg/toggle"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type PolicyReportEraser interface {
+	CleanupReportChangeRequests(ns string) error
+	EraseResultEntries(ns *string) error
+}
+
+func NewPolicyReportEraser(
+	pclient versioned.Interface,
+	reportLister policyreportv1alpha2listers.PolicyReportLister,
+	clusterReportLister policyreportv1alpha2listers.ClusterPolicyReportLister,
+) PolicyReportEraser {
+	return &eraser{
+		pclient:             pclient,
+		reportLister:        reportLister,
+		clusterReportLister: clusterReportLister,
+		mutex:               &sync.RWMutex{},
+	}
+}
+
+type eraser struct {
+	pclient             versioned.Interface
+	reportLister        policyreportv1alpha2listers.PolicyReportLister
+	clusterReportLister policyreportv1alpha2listers.ClusterPolicyReportLister
+	mutex               *sync.RWMutex
+}
+
+// TODO: make sure both functions below can be synced separately
+func (e *eraser) CleanupReportChangeRequests(ns string) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	var selector map[string]string
+	if ns != "" {
+		selector = map[string]string{ResourceLabelNamespace: ns}
+	}
+	return e.cleanupReportChangeRequests(selector)
+}
+
+func (e *eraser) EraseResultEntries(ns *string) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	return e.eraseResultEntries(ns)
+}
+
+func (e *eraser) cleanupReportChangeRequests(nslabels map[string]string) error {
+	var errors []string
+	var gracePeriod int64 = 0
+	deleteOptions := metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod}
+
+	selector := labels.SelectorFromSet(labels.Set(nslabels))
+
+	err := e.pclient.KyvernoV1alpha2().ClusterReportChangeRequests().DeleteCollection(context.TODO(), deleteOptions, metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		errors = append(errors, err.Error())
+	}
+
+	err = e.pclient.KyvernoV1alpha2().ReportChangeRequests(config.KyvernoNamespace()).DeleteCollection(context.TODO(), deleteOptions, metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		errors = append(errors, err.Error())
+	}
+
+	if len(errors) == 0 {
+		return nil
+	}
+
+	return fmt.Errorf("%v", strings.Join(errors, ";"))
+}
+
+func (e *eraser) eraseResultEntries(ns *string) error {
+	selector, err := metav1.LabelSelectorAsSelector(LabelSelector)
+	if err != nil {
+		return fmt.Errorf("failed to erase results entries %v", err)
+	}
+
+	var errors []string
+	var polrName string
+
+	if ns != nil {
+		if toggle.SplitPolicyReport.Enabled() {
+			err = e.eraseSplitResultEntries(ns, selector)
+			if err != nil {
+				errors = append(errors, fmt.Sprintf("%v", err))
+			}
+		} else {
+			polrName = GeneratePolicyReportName(*ns, "")
+			if polrName != "" {
+				polr, err := e.reportLister.PolicyReports(*ns).Get(polrName)
+				if err != nil {
+					return fmt.Errorf("failed to erase results entries for PolicyReport %s: %v", polrName, err)
```

Comment:
consider `errors.Wrapf` for all usages of `fmt.Errorf`

Reply:
fixed